notebooks/24-anomaly_detection.ipynb

```diff
@@ -0,0 +1,816 @@
+    "deletable": true,
+    "editable": true
+   },
+   "outputs": [],
+   "source": [
+    "kde_X = kde.score_samples(X)\n",
+    "print(kde_X.shape)  # contains the log-likelihood of the data. The smaller it is the rarer is the sample"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false,
+    "deletable": true,
+    "editable": true
+   },
+   "outputs": [],
+   "source": [
+    "from scipy.stats.mstats import mquantiles\n",
+    "alpha_set = 0.95\n",
+    "tau_kde = mquantiles(kde_X, 1. - alpha_set)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false,
+    "deletable": true,
+    "editable": true
+   },
+   "outputs": [],
+   "source": [
+    "n_samples, n_features = X.shape\n",
+    "X_range = np.zeros((n_features, 2))\n",
+    "X_range[:, 0] = np.min(X, axis=0) - 1.\n",
+    "X_range[:, 1] = np.max(X, axis=0) + 1.\n",
+    "\n",
+    "h = 0.1  # step size of the mesh\n",
+    "x_min, x_max = X_range[0]\n",
+    "y_min, y_max = X_range[1]\n",
+    "xx, yy = np.meshgrid(np.arange(x_min, x_max, h),\n",
+    "                     np.arange(y_min, y_max, h))\n",
+    "\n",
+    "grid = np.c_[xx.ravel(), yy.ravel()]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false,
+    "deletable": true,
+    "editable": true,
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "Z_kde = kde.score_samples(grid)\n",
+    "Z_kde = Z_kde.reshape(xx.shape)\n",
+    "\n",
+    "plt.figure()\n",
+    "c_0 = plt.contour(xx, yy, Z_kde, levels=tau_kde, colors='red', linewidths=3)\n",
+    "plt.clabel(c_0, inline=1, fontsize=15, fmt={tau_kde[0]: str(alpha_set)})\n",
+    "plt.scatter(X[:, 0], X[:, 1])\n",
+    "plt.legend()\n",
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "deletable": true,
+    "editable": true
+   },
+   "source": [
+    "## now with One-Class SVM"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The problem of density based estimation is that they tend to become inefficient when the dimensionality of the data increase. It's the so-called curse of dimensionality that affects particularly density estimation algorithms. The one-class SVM algorithm can be used in such cases."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true,
+    "deletable": true,
+    "editable": true
+   },
+   "outputs": [],
+   "source": [
+    "from sklearn.svm import OneClassSVM"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false,
+    "deletable": true,
+    "editable": true
+   },
+   "outputs": [],
+   "source": [
+    "nu = 0.05  # theory says it should be an upper bound of the fraction of outliers\n",
+    "ocsvm = OneClassSVM(kernel='rbf', gamma=0.05, nu=nu)\n",
```

Comment:
gamma was set so it looked nice. but exercise raise the problem of hyperparameter setting